src/CommunityToolkit.Maui.SourceGenerators/Generators/TextColorToGenerator.cs

```diff
@@ -18,97 +26,80 @@ class TextColorToGenerator : IIncrementalGenerator
+			static (compilation, token) =>
 			{
-				var diag = Diagnostic.Create(TextColorToDiagnostic.InvalidClassDeclarationSyntax, Location.None, classDeclarationSyntax.Identifier.Text);
-				context.ReportDiagnostic(diag);
-				continue;
-			}
-
-			// If the control is inherit from a Maui control that implements ITextStyle
-			// We don't need to generate a extension method for it.
-			// We just generate a method if the Control is a new implementation of ITextStyle and IAnimatable
-			var doesContainSymbolBaseType = mauiTextStyleImplementors.ContainsSymbolBaseType(declarationSymbol);
+				var iTextStyleInterfaceSymbol = compilation.GetTypeByMetadataName(iTextStyleInterface);
+				var iAnimatableInterfaceSymbol = compilation.GetTypeByMetadataName(iAnimatableInterface);
 
-			if (!doesContainSymbolBaseType
-				&& declarationSymbol.AllInterfaces.Contains(textStyleSymbol, SymbolEqualityComparer.Default)
-				&& declarationSymbol.AllInterfaces.Contains(iAnimatableSymbol, SymbolEqualityComparer.Default))
-			{
-				if (declarationSymbol.ContainingNamespace.IsGlobalNamespace)
+				if (iTextStyleInterfaceSymbol is null || iAnimatableInterfaceSymbol is null)
 				{
-					var diag = Diagnostic.Create(TextColorToDiagnostic.GlobalNamespace, Location.None, declarationSymbol.Name);
-					context.ReportDiagnostic(diag);
-					continue;
+					throw new Exception("There's no .NET MAUI referenced in the project.");
 				}
 
-				var nameSpace = declarationSymbol.ContainingNamespace.ToDisplayString();
+				var mauiAssembly = compilation.SourceModule.ReferencedAssemblySymbols.Single(q => q.Name == mauiControlsAssembly);
+				var symbols = GetMauiInterfaceImplementors(mauiAssembly, iAnimatableInterfaceSymbol, iTextStyleInterfaceSymbol).Where(x => x is not null);
 
-				var accessModifier = GetClassAccessModifier(declarationSymbol);
+				return symbols;
+			});
 
-				if (accessModifier == string.Empty)
-				{
-					var diag = Diagnostic.Create(TextColorToDiagnostic.InvalidModifierAccess, Location.None, declarationSymbol.Name);
-					context.ReportDiagnostic(diag);
-					continue;
-				}
 
-				textStyleClassList.Add((declarationSymbol.Name, accessModifier, nameSpace, declarationSymbol.TypeArguments.GetGenericTypeArgumentsString(), declarationSymbol.GetGenericTypeConstraintsAsString()));
-			}
-		}
+		// Here we Collect all the Classes candidates from the first pipeline
+		// Then we merge then with the Maui.Controls that implements the desired interfaces
+		// Then we make sure they are unique and the user control doesn't inherit from any Maui control that implements the desired interface already
+		// Then we transform the ISymbol to be a type that we can compare and preserve the Incremental behavior of this Source Generator
+		var inputs = userGeneratedClassesProvider.Collect()
+			.Combine(mauiControlsAssemblySymbolProvider)
+			.SelectMany(static (x, _) => Deduplicate(x.Left!, x.Right).ToImmutableArray())
```

Comment:
I'm using the `!` here, because for some reason the roslyn didn't see that I filtered for not null values, at line `62`

Reply:
I'll keep poking around to try and find another way to avoid `!`. But if I can't figure out something by the end of the day today, we can merge it and create a Bug Report noting it so that we don't forget.

Perhaps we could ping the Roslyn team on this to add some [Nullable static analyzers](https://learn.microsoft.com/dotnet/csharp/language-reference/attributes/nullable-analysis) to help other devs avoid this too.

Reply:
Fixed. 

I just passed through the nullable `INamedTypeSymbol?` to the `Deduplicate()` method, and do the null check there.

```cs
static IEnumerable<INamedTypeSymbol> Deduplicate(ImmutableArray<INamedTypeSymbol?> left, IEnumerable<INamedTypeSymbol> right)
{
  foreach (var leftItem in left)
  {
	  if (leftItem is null)
	  {
		  continue;
	  }
      // ...
}
```